### PR TITLE
add type annotations to torch.nn.modules.container

### LINKED ISSUE #48968: Enable torch.nn.modules.container typechecks during CI
`TestTypeHints.test_run_mypy` should pass with `[mypy-torch.nn.modules.container]` deleted from mypy.ini

cc @ezyang @malfet @rgommers @xuzhao9 @gramster

### PR DESCRIPTION
**Pull request by guilhermeleobas**
Fixes #48968

### CODE REVIEW

File: torch/nn/modules/container.py
Comment (rgommers) [inline torch/nn/modules/container.py:374]: Agreed with this ignore. Would be good to add a comment above it. Something like:
```
# modules can be Mapping (what it's typed at), or a list: [(name1, module1), (name2, module2)]
# that's too cumbersome to type correctly with overloads, so we add an ignore here
```

File: torch/nn/modules/container.py
Comment (rgommers) [inline torch/nn/modules/container.py:645]: Same here, brief comment `# parameters as length-2 list too cumbersome to type, see ModuleDict.update comment`

**Review by rgommers — APPROVED** [review]
LGTM, thanks @guilhermeleobas. Just adding a couple of comments would be nice.

**Comment by guilhermeleobas** [comment]
Hi @rgommers, done!

**Review by rgommers — APPROVED** [review]
LGTM now

**Review by walterddr — APPROVED** [review]
looks good. thanks for the contribution!

**Comment by guilhermeleobas** [comment]
Do not merge this PR until one checks if the annotations introduce any regression. See:
https://github.com/pytorch/pytorch/pull/49564#issuecomment-752916072

**Review by walterddr — COMMENTED** [review]
looks like the only thing we need to add as additional test are
```
model = ModuleList(...) 
model += nn.Conv1d() 
model.append(nn.Conv2d()) 
model.extend([nn.Conv1d(), nn.Conv2d())
```
doesn't have to be convolution modules. we just need to validate that the __iadd__, append, extend APIs are ok with JIT

**Comment by guilhermeleobas** [comment]
@walterddr there is a test on `test_nn.py` called `test_ModuleList` which asserts these properties.

**Comment by guilhermeleobas** [comment]
Done! @walterddr 

Should I include tests for `ParameterList` and `Sequential` as well?

**Comment by walterddr** [comment]
yes good catch. for `ParameterList` it would be good if we add them. `Seqeuential` should be fine since you only annotated those dunder methods and they should already be covered by existing jit test.

**Comment by guilhermeleobas** [comment]
I am having some issues to add `ParameterList`:
```python
class Param(nn.Module):
    def __init__(self):
        super(Param, self).__init__()
        self.params = nn.ParameterList([nn.Parameter(torch.randn(10, 10)) for i in range(10)])
        # p = nn.Parameter(torch.randn(10, 10))
        # self.params = nn.ParameterList([p for _ in range(10)])
        # self.params += (p,)
        # self.params.append(p)
        # self.params.extend([p, p, p])

    def forward(self, x):
        # ParameterList can act as an iterable, or be indexed using ints
        # for i, p in enumerate(self.params):
        sz = len(self.params)
        for i in range(sz):
            p = self.params[i]
            x = self.params[i // 2].mm(x) + p.mm(x)
        return x

m = torch.jit.script(Param())
```

And the error:
```
    def create_methods_and_properties_from_stubs(concrete_type, method_stubs, property_stubs):
        method_defs = [m.def_ for m in method_stubs]
        method_rcbs = [m.resolution_callback for m in method_stubs]
        method_defaults = [get_default_args(m.original_method) for m in method_stubs]

        property_defs = [p.def_ for p in property_stubs]
        property_rcbs = [p.resolution_callback for p in property_stubs]

>       concrete_type._create_methods_and_properties(property_defs, property_rcbs, method_defs, method_rcbs, method_defaults)
E       RuntimeError:
E       Tried to access nonexistent attribute or method '__len__' of type '__torch__.torch.nn.modules.container.ParameterList'. Did you forget to initialize an attribute in __init__()?:
E         File "/home/guilhermel/git/pytorch/test/test_jit.py", line 848
E                       # ParameterList can act as an iterable, or be indexed using ints
E                       # for i, p in enumerate(self.params):
E                       sz = len(self.params)
E                            ~~~ <--- HERE
E                       for i in range(sz):
E                           p = self.params[i]
```

**Comment by guilhermeleobas** [comment]
Last commit introduce tests for `Sequential`. I couldn't find a way to create a test for `ParameterList`.

**Comment by walterddr** [comment]
that should be fine consider the annotations on ParameterList are essentially the same ones on ModuleList

**Comment by walterddr** [comment]
test `test_ParameterList` in test_nn.py actually failed for this. so yes I guess ParameterList is covered and didn't quite work. could you take a look?

**Comment by guilhermeleobas** [comment]
I had commit some changes by mistake in the `__getitem__` function. 

```python
================================================================== test session starts ==================================================================
platform linux -- Python 3.9.1, pytest-6.2.1, py-1.10.0, pluggy-0.13.1 -- /home/guilhermel/.conda/envs/pytorch-cuda-dev/bin/python
cachedir: .pytest_cache
hypothesis profile 'default' -> database=DirectoryBasedExampleDatabase('/home/guilhermel/pytorch/.hypothesis/examples')
rootdir: /home/guilhermel/pytorch
plugins: hypothesis-6.0.0
collected 2054 items / 2051 deselected / 3 selected

test/test_nn.py::TestNN::test_ParameterList PASSED                                                                                                [ 33%]
test/test_nn.py::TestNN::test_parameterlistdict_pickle PASSED                                                                                     [ 66%]
test/test_nn.py::TestNN::test_parameterlistdict_setting_attributes PASSED                                                                         [100%]
```

Tests for `ParameterList` on `test_nn.py` are passing now.

**Review by malfet — APPROVED** [review]

**Review by walterddr — APPROVED** [review]

**Comment by walterddr** [comment]
could you fix the test failures? seems like Parameter needs to be imported.

**Comment by guilhermeleobas** [comment]
Looks like CI is all green

**Comment by guilhermeleobas** [comment]
Fixed CI issue in my last commit

**Pull request closed without merging**
